### PR TITLE
Use new string xplat APIs

### PR DESCRIPTION
**Pull request by agarwal-sandeep**
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

chakrashim

##### Description of change

Merge chakracore-master into xplat branch
Sync to latest Microsoft/ChakraCore for new string APIs
Fix a bug in ChakraCore which is in PR Microsoft/ChakraCore/pull/1988
Fix build break for new string APIs

### CODE REVIEW

**Comment by digitalinfinity** [comment]
Reviewed a53e51c and it looks good to me. Couple small comments:

File:.\deps\chakrashim\src\jsrtutils.cc, Total comments:1
Comment #1, L1007 of right file<1019>
Reviewer comment:
    Should check for malloc succeeding here?
Source line:
      uint8_t\* buffer = (uint8_t*)malloc(len+1);

File:.\deps\chakrashim\src\v8string.cc, Total comments:1
Comment #1, L34 of right file<34>
Reviewer comment:
    Should check for malloc succeeding here?
Source line:
      uint8_t\* buffer = (uint8_t*)malloc(len + 1);

**Comment by kunalspathak** [comment]
Can you please fix the title of commits as per contribution guidelines?

**Comment by kunalspathak** [comment]
f2323e6b1a5c088cbad0139a8c64a9935fb13a5d LGTM

File: deps/chakrashim/src/jsrtdebug.cc [hunk 1]
```diff
@@ -40,15 +40,18 @@ namespace jsrt {
 
   v8::Local<v8::String> MessageImpl::JsValueRefToV8String(JsValueRef result) {
```
Comment (jianchun) [inline deps/chakrashim/src/jsrtdebug.cc]: Not sure why this is needed... in chakrashim we can mostly cast a JsValueRef to v8::Local directly. So this function can simply first ensure the JsValueRef is a String (maybe there is a helper do the conversion), then a cast like `Local<String>::New(result...)` would suffice.

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 2]
```diff
@@ -377,4 +380,26 @@ void* IsolateShim::GetData(uint32_t slot) {
   return slot < _countof(this->embeddedData) ? embeddedData[slot] : nullptr;
 }
 
+JsValueRef IsolateShim::GetChakraShimJsArrayBuffer() {
+  if (this->chakraShimArrayBuffer == nullptr) {
+    CHAKRA_VERIFY(JsCreateExternalArrayBuffer(
+                  (void*)jsrt::chakra_shim_native,
+                  sizeof(jsrt::chakra_shim_native),
+                  nullptr, nullptr,
+                  &this->chakraShimArrayBuffer) == JsNoError);
```
Comment (jianchun) [inline deps/chakrashim/src/jsrtisolateshim.cc]: Not sure if these 2 jslib sources are needed multiple times. They are mostly only needed at initialization. No need to save/cache them in IsolateShim (also their lifetime may be a bit problematic, no explicit code to keep these cached Vars alive).

File: deps/chakrashim/src/v8signature.cc [hunk 3]
```diff
@@ -53,7 +53,7 @@ bool Utils::CheckSignature(Local<FunctionTemplate> receiver,
   if (!matched) {
     const char txt[] = "Illegal invocation";
     JsValueRef msg, err;
-    if (JsPointerToStringUtf8(txt, _countof(txt) - 1, &msg) == JsNoError &&
+    if (JsCreateString(txt, _countof(txt) - 1, &msg) == JsNoError &&
```
Comment (jianchun) [inline deps/chakrashim/src/v8signature.cc]: Here it is `_countof() - 1`... Please check the jslib sources. Should we use `_countof() - 1` there, or `sizeof() - 1` or `sizeof()` (currently)?

File: deps/chakrashim/src/jsrtcontextshim.cc [hunk 4]
```diff
@@ -450,22 +447,13 @@ bool ContextShim::ExecuteChakraShimJS() {
...
-    return false;
-  }
-
-  // Ensure the buffer is null terminated
-  buffer[_countof(chakra_debug_native)] = L'\0';
-
   JsValueRef getInitFunction;
-  if (JsParseScriptWithAttributes(buffer,
+  JsValueRef url;
+  CHAKRA_VERIFY(JsCreateString("chakra_debug.js", strlen("chakra_debug.js"),
```
Comment (jianchun) [inline deps/chakrashim/src/jsrtcontextshim.cc]: It looks a template helper for JsCreateString and JsCreatePropertyId that automatically deduce C string literal length would reduce a lot of "strlen".

File: deps/chakrashim/src/v8signature.cc (hunk 3 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8signature.cc]: Used sizeof instead to be consistent.

File: deps/chakrashim/src/jsrtcontextshim.cc (hunk 4 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/jsrtcontextshim.cc]: Good suggestion, created utility functions jsrt::CreateString and jsrt::CreatePropertyId

**Comment by jianchun** [comment]
LGTM, thanks!

**Pull request merged**